### DISCUSSION TITLE
TemporaryUploadedFile serialization issue

### DISCUSSION THREAD

**Discussion by gaetan-hexadog** [livewire/livewire#3134] [7. Bugs (archived)]
### Description
When trying to pass a TemporaryUploadedFile to an event (to pass it to a parent component), the class is not well serialized:

Following object passed into an `emit` call
```php
Livewire\TemporaryUploadedFile {#1659 ▼
    +"disk": "local"
    #storage: Illuminate\Filesystem\FilesystemAdapter {#1893 ▶}
    #path: "livewire-tmp/eypmQmwEA2I438gXJCvNnRUwY0yXOB-metadGltZS1tYW5hZ21lbnQucG5n-.png"
    -test: false
    -originalName: "eypmQmwEA2I438gXJCvNnRUwY0yXOB-metadGltZS1tYW5hZ21lbnQucG5n-.png"
    -mimeType: "application/octet-stream"
    -error: 0
    #hashName: null
    path: "/private/var/folders/vv/6gx44q_s3_q2pws4hwhxdcs00000gn/T"
    filename: "eypmQmwEA2I438gXJCvNnRUwY0yXOB-metadGltZS1tYW5hZ21lbnQucG5n-.png"
    basename: "phpqHUhwY"
    pathname: "/private/var/folders/vv/6gx44q_s3_q2pws4hwhxdcs00000gn/T/phpqHUhwY"
    extension: ""
    realPath: "/Users/gaetan/Desktop/Projets/laravel/storage/app/livewire-tmp/eypmQmwEA2I438gXJCvNnRUwY0yXOB-metadGltZS1tYW5hZ21lbnQucG5n-.png"
    aTime: 2020-12-11 20:41:11
    mTime: 2020-12-11 20:41:11
    cTime: 2020-12-11 20:41:11
    inode: 3616610
    size: 125358
    writable: false
    readable: false
    executable: false
    file: false
    dir: false
    link: false
  }
```

is retreived like that at the end:

```php
array:1 [▼
    "disk" => "local"
  ]
```

only `+"disk": "local"` property is serialized. Not the all object. So I can't access to the TemporaryUploadedFile into listeners.

**Context**
- Livewire version: 2.32
- Laravel version: 8.13

**Comment by MikeGlennFWC** [comment]
Spent hours trying to figure this out.

For anyone else having this issue, I can only imagine you are trying to pass your upload(s) from one component to another. Unfortunately the `emit` methods do not automatically serialize `TemporaryUploadedFile` objects, unlike when a component is `hydrating` or `dehydrating`.

What we did to combat this was, using the `updated` lifecycle hook, we manually serialized the file(s) before calling emit, in the receiving component event function, we then unserialized. 

I believe when the Event system is serializing and deserializing, it should perhaps be checking to see if there are any `TemporaryUploadedFile` objects and serialize them into the `livewire-file(s)`

@calebporzio what are your thoughts here? Do you think this is a reasonable assumption or is there something which would cause you to advise against it?


See below some example code of how we got this to work.

in the FileUpload component, updated lifecycle hook.

```php
    public $files = [];
    ...
    public function updatedFiles($value)
    {
        $serializedFiles = TemporaryUploadedFile::serializeMultipleForLivewireResponse($value);
        $this->emitUp('documentUploaded', $serializedFiles);
    }
```

In the receiving component:
```php
    public function documentUploaded($value)
    {
        if (TemporaryUploadedFile::canUnserialize($value)) {
            $this->files = TemporaryUploadedFile::unserializeFromLivewireRequest($value);
        }
    }
```

**Comment by kornatzky** [comment]
I fixed this by serializing and storing it in the session

**Reply by fanorama** [reply]
Hey i had into this problem, how do you solved it? how to store in the session?

**Reply by kornatzky** [reply]
While I thought I solved it, I did not. So, I had to retract using the livewire wizard package from Spatie. This is how the problem originated. Now there is Livewire 3 so it may have resolved the problem. But I did not try Livewire 3 yet.

**Comment by ahmedsayedabdelsalam** [comment]
is there a fix for that ? i also have uploader component that need to emit uploaded temp files to parent component to save the images with the model

**Reply by fanorama** [reply]
have you solved it?

**Reply by ahmedsayedabdelsalam** [reply]
with a workaround unfortunately
